### PR TITLE
Add units of measure

### LINKED ISSUE #32: Consider Units of Measure when getting or setting Socket, Context, or Message Options
`Socket`, `Context`, and `Message` can get and/or set various options, each of which is an `int` in native code. Perhaps Units of Measure can be used to improve the experience when working with this portion of the API?

### PR DESCRIPTION
**Pull request by DanielFabian**
fixes #32. Please review it, if it's about what you were thinking.

### CODE REVIEW

**Comment by pblasucci** [comment]
This is very interesting. Although it wasn't what I was expecting. I'm really sorry I didn't communicate it  better. I was mostly fine with many of the constants and error codes not having units. However, the `Context`, `Message`, and `Socket` module all have `getOption`/`setOption` pairs. Currently, these all take a constant (identifying the option). Further, the setters take a primitive value and the getters return a primitive value. It is to these primitive values that I was hoping to attach units (at least for the numeric primitives). My hope was that I could capture the units of the various settings without resorting to a speific getter/setter pair for each option. Perhaps a brief example will clarify...

Currently, we have something like this:
```
Socket.setOption mySocket (ZMQ. MULTICAST_HOPS,5)
Socket.setOption mySocket (ZMQ. RCVTIMEO,5)
```
Both calls take a constant (an `Int32`) and a value (also an `Int32`). However, the first `5` is network hops, while the other `5` is milliseconds. This is only captured in documentation... but it'd be great to capture it in the type system.

What I'd like to be able to write (in an ideal world):
```
Socket.configure mySocket [ (ZMQ.MULTICAST_HOPS, 5<network_hops>) 
                            (ZMQ.RCVTIMEO      , 5<milliseconds>) ]
```
Of course, I realize this might not be 100% achievable (Heck! It might not even be advisable). But something along these lines was more what I was thinking. Although, I must say, I really do like the idea of `SocketType` as unit-of-measure. 

Anyway, I'm not sure where to go from here. But I really appreciate your efforts, and would really enjoy having you continue to contribute. Let me know how you'd like to proceed (also, if you see me on Slack, feel free to continue the discussion there, too).

**Comment by DanielFabian** [comment]
Your original idea where you'd make a tuple with the constant and a value with units of measure is not going to work, as you would have different types per list member. This new idea, however, where you use a function with typed arguments might just work out.

We'd need to make functions like 

```
val MulticastTTL : int<network_hops -> int<SocketOption> * int
val RecvTimeout : int<milliseconds> -> int<SocketOption> * int
```

Then `Socket.configure` can simply iterate over the list and apply each `int<SocketOption> * int` pair to `mySocket`. This should not be hard. I'll try messing around with it.

The consequence is, though, that we'll need a function for each `SocketOption` constant, that can be meaningfully set to some value.

In this context, the version of `setOption` is sort of the *untyped* version of setting them and version with a specific version for each constant is more like the *typed* version.

As for getting the option, i.e. reading the value, I'm not perfectly sure just yet, what this ought to look like; possibly `getOption` would become a higher-order-function?

```
val getOption : (int<'a> -> int<SocketOption> * int) -> int<'a>
```

We would then pass, say, `MulticastTTL` to `getOption`. Internally call the function with `LanguagePrimitives.IntegerWithMeasure 0` to get the constant. Then actually get the option and then return the value, say `opt` as `LanguagePrimitives.IntergerWithMeasure opt`. With this, the type system should thread through the types we need. 

I'll mess around with that, too.

I'll keep you posted ;-)

**Comment by pblasucci** [comment]
Oooo... this is all very interesting. And it's caused lots of ideas to roll around my nearly-empty head. 

:-)

I was originally hoping to avoid having to define anything on a per-property basis. But that's probably unavoidable if we want a better API. That having been said, I like *a lot* of what you've outlined above. I do think we ought to drop the `SocketOption` unit-of-measure. This would mean `setOption` and `getOption` have no change. However, we could then add the more-strongly-typed converter layer on top. Also, in terms of the new layer, functions are cool... but what are your thoughts on using a discriminated union?

Additionally, I'd very much like to pull in the `SocketType` unit-of-measure, as I think it fits perfectly. And it's overall impact on existing consumers is negligible. However, I think the other units (errors, socket options, et cetera) are probably going to cause more headaches than they solve.

Feedback wanted.

**Comment by DanielFabian** [comment]
If you're not in a hurry, I'd like to let it cook for a few iterations,
whilst playing around with different APIs, like functions or DUs and when
we get to the point, that we actually want one particular option, it's not
going to be hard to create a branch with the wanted changes.

**Comment by pblasucci** [comment]
Sure. There's no rush at all on this. In fact, I'd rather it take a little longer and be really awesome. I just didn't want you to associate a lack of merges with a lack of interest.

**Comment by DanielFabian** [comment]
Please look at the commit https://github.com/DanielFabian/fszmq/commit/0707579da3d903d3b3e860d57ffc5d301bae414b and the comment over there

**Comment by DanielFabian** [comment]
I was playing around with active patterns. Here, I use non-partial active patterns to read out options. 

And with some tricks as to where one puts them, one can actually get "overloading" of the DU-pattern case and the active pattern to work. Though, I am not so sure, whether this is helpful or more like confusing.

Maybe something like `SocketOption.Rate` and maybe a module `GetSocketOption` for the active patterns would be better?

Then you could write something like 
```
let (GetSocketOption.Rate rate) = socket
printfn "%d" rate
```
or 
```
socket |> fun (GetSocketOption.Rate rate) -> printfn "%d" rate
```

Feedback more than welcome

**Comment by pblasucci** [comment]
@DanielFabian I like the idea of Active Patterns for reading the options. It'd probably be sufficient to  have a library of Partial Active Patterns, which all use the existing `getOption` internally,  which all take in  a `Socket` (or `Context` or `Message`) instance and return the appropriate value (with Units of Measure, where appropriate). The Active Patterns combined with the Discriminated Union approach for setting options feels like a really good way to go about things.

**Comment by DanielFabian** [comment]
What would be the partial case though? Does `getOption` throw?

**Comment by pblasucci** [comment]
Well, technically, `getOption` can throw in a couple of different situations. Specifically:

1. You pass it a bad input
2. You're environment failed to resolve/load `libzmq` correctly

But the reason I was thinking of Partial Active Patterns was about composability. Not that Total Active Patterns would be wrong... just that the composability semantics would be different.

**Comment by DanielFabian** [comment]
Could you please make an example? I don’t think, I understand what you mean
by this.

In particular, when you pattern match against the socket with say Rate, the
active pattern will return you the rate. And I do not quite understand what
would be the None case. I mean getOption will always return the rate, if
you ask for it and throw, otherwise.

The way I see it, the active pattern is not composable anyway, much in the
same way, like, say, tuple decomposition.

Mind you, there are three types of active patterns; total, partial and
choice patterns. Total transform your input, partial check, whether your
input has a certain *shape* and choice patterns partition you domain into
certain cases.

Unless is misunderstood it, and it is possible, that in certain states of
the socket you maybe cannot ask for the rate or something?
​

**Comment by pblasucci** [comment]
In terms of composability, I was thinking about getting multiple options simultaneously (though that may not be the best idea).  But forget what I was saying about partial patterns... I wasn't thinking too clearly this morning. You are correct that partial patterns don't make sense here.

**Comment by DanielFabian** [comment]
Ok, it is settled then. I'll refactor it towards that. But it'll take a few days, possibly weekend, but maybe even a little longer

**Comment by pblasucci** [comment]
Awesome. No rush at all. However, I had a thought about organization (and backward compatibility). Specifically, let's leave `setOption` and `getOption` unchanged. Additionally, let's put all the cool new stuff in it's own file (maybe call it `Config.fs` or `Options.fs`). I'm thinking some auto-open sub-modules (`Socket`, et cetera) under a manually opened top-level module (named similarly to the file) might make for a reasonably good experience.

**Comment by DanielFabian** [comment]
I added the changes as we discussed. But there are a few (marked as todos) that i can't guess how to use. Could you read the code? And also, how can we test it?

**Comment by pblasucci** [comment]
Thanks! This looks really great. 

I'm not certain I understand your concerns about the `TCP_KEEPALIVE`* options (but I _do_ need to better document them). At any rate, I suspect we can leave these "as is" for now. We can always tweak them later if we need to.

As for testing, that's an interesting question. Perhaps we can use the `Socket.getOption` and `Socket.setOption` as "test oracles"? In other words, use the older functions to verify the behavior of the new functions.

**Comment by pblasucci** [comment]
I've been thinking about a potential wrinkle, with respect use using total patterns. As we roll out support for new versions of `libzmq`, the options will change. Additionally, the valid values, both for getters and setters, will changes. In both cases, but especially for the getters, this is going to cause breaking changes in terms I type signatures (_I should mention I'm trying very hard to ensure that one version of `fszmq` can be used with many versions on `libzmq`).

I'm not sure why I hadn't thought of this earlier... :-(   

I wonder if this makes the whole thing somewhat less palatable.

Ideas (eagerly) wanted.

**Comment by DanielFabian** [comment]
Well, there are really two things to consider here. If you extend the
setters, then that should not hurt at all, because you don’t ever need to
pattern match against that (other than zmq itself, which will, of course,
be extended). On the other hand, extending the total patterns should not
hurt either, for they are just functions but you cannot combine them in a
pattern match anyway.

So any change should be source code compatible between versions (modulo
options that are completely dropped, as they would fail).

The only thing I am not sure as to whether or not it works out-of-the-box
is a binding redirect. I mean fully qualified name of a type can be binding
redirected to a newer version which should in principle remain compatible,
so the question is, what happens in case we have an instance of a type that
is say Case1 in a DU MyDU. Then in IL, Case1 is a class derived from MyDU.
In principle, that should also work after a binding redirect. So the
question is, what happens if we add more cases, but IL-wise a new Case2
should just be an additional class which also derives from MyDU. Which
again should work, I think.

When I look at the things, so far, I don’t see breaking changes, especially
not for new getters, as they already are just fully independent functions,
being separate active patterns for each case. And the setters would only
become breaking changes, if you were to pattern match against the DU, which
you shouldn’t be doing anyway, as it is fairly pointless; but there is
always the option of just providing factory methods corresponding to the DU
cases and hiding the DU implementation, that way an instance of MyDU can
only ever be created using the factory methods but not pattern matched
against (though I think this is an overkill)
​

**Comment by pblasucci** [comment]
As usual, excellent points. 

And I agree, setters aren't likely a big deal. 

However, there is one troubling issue for the getters -- multiple-case total patterns. At the moment, I believe this is only used for security mechanism. But that's exactly the option about which I'm worried. I'm mostly done with a patch to support version 4.1.3 of `libzmq`. However, it introduces a new security mechanism. This means the current pattern `(|NoSecurity|Plain|Curve|)` needs to become `(|NoSecurity|Plain|Curve|GSSAPI|)`. Now, in practice, because we've not yet released _any_ of these changes, we can merge things together and we'll be fine. _(As an aside, we'll need to have bit of co-ordination between the release of your work and the patch I'm finishing up.)_ But, the whole situation got me thinking... we'll potentially hit this issue at some point in the future; so we should probably come up with a plan for it now.

**Comment by DanielFabian** [comment]
Good idea, I'll think about it a little and come back to you in a while, if
that is ok.

**Comment by DanielFabian** [comment]
To keep backwards compatibility, we could make partial active patterns for
cases that constitute enum values like type of security: (|NoSecurity|_|),
(|Plain|_|), (|Curve|_|) and then we can later add (|GSSAPI|_|) without
losing backward compatibility.

This does not give us a 100% guarantee that nothing will ever cause
breaking changes, but it should hep a lot. In particular, things like e.g.
getting the rate will remain backwards compatible, until 0MQ comes up with
the idea to return the value in MB now instead of Bytes or if they change
to int64.

So with this, we remain open for extension and only incur breaking changes
upon semantic changes to 0MQ itself. I feel, that this ought to be an
acceptable trade-off.
​

**Comment by pblasucci** [comment]
Yeah... I'm digging it. I think you're right. Partial patterns for "enum-like" properties is a good compromise. 

Also, I don't know if you saw, but there was recently a patch to the ZMQ docs to better explain some of the `TCP_*` properties of which you were uncertain.

**Comment by pblasucci** [comment]
Hi @DanielFabian ... So, we've got a bit of scheduling to sort out. I (finally) have a patch ready which will introduce support for 4.1.x (while preserving existing 4.0.x support). However, I wasn't sure -- from an effort perspective -- if made sense to incorporate your changes first. If we bring in my patch first, yours won't have complete option coverage. However, that might be OK (we could merge a partial with partial support into master... but I would bump the version and put out a new release of the NuGet until the support was complete). Alternatively, if we bring in _your_ patch first, it would be complete. So we could do a version bump and a NuGet release (wherein the new Socket options would be the last pre-4.1.x feature). Either way, at some point, the new options added by my patch need to be exposed via the changes you are introducing. Since I can very much see arguments in favor of either approach, I'd love to hear your thoughts.

**Comment by DanielFabian** [comment]
for me, it sounds like a better idea to do it after the release. Because as
of this moment, there are no unit tests and they should be definitely
written before releasing it.

**Comment by pblasucci** [comment]
Makes sense. In that case, I'm going to try and get a release out later today.

**Comment by haf** [comment]
This code looks really nice; also helps a lot in API discoverability. Is a release of this coming?

**Comment by pblasucci** [comment]
Funnily enough, Daniel and I were just corresponding about this this morning. The short answer is I plan to get it integrated in the next week or so... but I don't have a firm date (too many other commitments).

**Comment by DanielFabian** [comment]
could you please sum up, what's the status around this? Turns out, I probably will have some time in the next week or so to finally finish it up.

I need to know what's still open, apart from my own TODOs.

**Comment by pblasucci** [comment]
Hi @DanielFabian 

I started working on integrating your existing patch, but haven't finished. 

:-(

Basically, I've been going through and adding the various getters and setters with appropriate documentation. However, it's slow-going because I often need to consult a mix of sources to figure out the right "options" to present. The `libzmq` docs are helpful but often insufficient. In fact, I find it's necessary to peek in the `libzmq` source and also do a bit of googling (especially when parts of ZeroMQ reuse "standard" networking terms, values). There are 11 options still outstanding for compatibility with libzmq 4.1.x. Also, I've only added one unit test. So there's quite a few of those to be done as well.

I'd be thrilled if you resumed this work. I'll commit the branch in a little while. And we can go from there (in case you have questions or whatever). Thanks!

Welcome back!

:-)

**Comment by pblasucci** [comment]
@DanielFabian the WIP is here: https://github.com/pblasucci/fszmq/tree/new-option-api

**Comment by DanielFabian** [comment]
great, I'll look at it tomorrow or the day after.

**Comment by DanielFabian** [comment]
merged your branch into the PR. I am working on the unit tests now. I should have something in a few hours.

**Comment by DanielFabian** [comment]
alright, so I added many unit tests. Turns out some options don't seem to support reading, for a few we'll need different unit tests, e.g. Subscribe/Unsubscripe and security. 

So far, nothing too scary.

But I found one, that I find problematic:

    AuthenticationDomain "my domain"

reads back as
    
    AuthenticationDomain "my domain\0"

i.e. the string that's read back is zero-terminated, whereas the .net one wasn't. Do we need to `TrimEnd '\000'` in `Marshal.fs`, line 41?

**Comment by pblasucci** [comment]
This is outstanding, thanks! As far as the `TrimEnd` in `Marshsall.fs` that makes perfect sense (and you're right... It's not uncommon in .NET interop code).

Super excited to see this coming together!  :-)

**Comment by DanielFabian** [comment]
Alright, I got most of the tests done for this. Two are still failing, but they shouldn't be too hard.

And then it's all about your 11 TODO cases, I believe.

**Comment by DanielFabian** [comment]
As a side-note, do you have any particular reason for not adding a curve-enabled libzmq.dll? I had to compile it to be able to write curve tests.

**Comment by DanielFabian** [comment]
Alright, now the tests are green. We need to test Subscribe / Unsubscribe and your TODOs.

**Comment by DanielFabian** [comment]
subscribe / unsubscribe done. They can't be tested on the getter/setter level though, I can only test setting them doesn't blow up. Next, I'll look at the TODOs.

**Comment by DanielFabian** [comment]
here's the remaining work:

 - [X] HANDSHAKE_IVL
 - [ ] LAST_ENDPOINT
 - [ ] ROUTER_HANDOVER
 - [ ] TOS
 - [ ] CONNECT_RID
 - [ ] GSSAPI_SERVER
 - [ ] GSSAPI_PRINCIPAL
 - [ ] GSSAPI_SERVICE_PRINCIPAL
 - [ ] GSSAPI_PLAINTEXT
 - [ ] SOCKS_PROXY
 - [ ] XPUB_NOD

**Comment by pblasucci** [comment]
Really digging how this is going... 

As for Curve support, my logic was to put the smallest, simplest `libzmq` into the repo. Especially since full testing requires changing the referenced library no matter what (i.e. If you normally use a copy _with_ Curve support, you have to temporarily swap it out to confirm that the "has capability" functions correctly return `false`). Admittedly, this is a bit of a coin-toss, and not something to which I really gave a lot of thought. We can change the version stored in the repo if it makes more sense for testing purposes. 
:-)

**Comment by DanielFabian** [comment]
good point, we need to:

 - [ ] make the tests green/yellow also with a non-curve libzmq
 - [ ] make sure, the security mechanism active patterns don't throw for a non-curve libzmq

**Comment by DanielFabian** [comment]
Alright, with my last commit, I think, this PR is concluded. As for the GSSAPI, I implemented the protocol according to the spec / C sources as far as I understood it. But unfortunately, I don't have a GSSAPI-enabled libzmq, so it would probably be prudent to run the unit tests on OS X or something, just to make sure they are green.

At any rate, I don't have anything planned to do left, at the moment. Please do review the code and possible merge it ;-)

**Comment by haf** [comment]
@DanielFabian if you tell me how to test GSSAPI I can test it

File: src/fszmq/Options.fs [hunk 1]
```diff
@@ -82,14 +82,16 @@ module Options =
...
+    | PlainServer
     /// Set user name and password for client using PLAIN security
     | PlainClient of username:string * password:string
     /// Make socket act as a server for CURVE security
-    | CurveServer of enabled:bool * secretKey:byte[]  
-    /// Public and private keys fo client using CURVE security
-    | CurveClient of publicKey:byte[] * secretKey:byte[] * serverPublicKey:string 
+    | CurveServer of secretKey:byte[]  
+    /// Public and private keys for client using CURVE security
+    | CurveClient of publicKey:byte[] * secretKey:byte[] * serverKey:byte[]
```
Comment (haf) [inline src/fszmq/Options.fs]: 🎆 💟

**Comment by DanielFabian** [comment]
Just build the solution and run the unit tests. The tests are actually all written and simply assert that GSSAPI is enabled. If it isn't the unit tests Assert.Ignore.

So build the solution and replace the libzmq in the build output before running the unit tests.

**Comment by haf** [comment]
I have libzmq globally installed, won't that work?

**Comment by pblasucci** [comment]
@haf It will if it's built with the correct options. But you might need to delete the local `libzmq` from `bin\Debug`.

**Comment by DanielFabian** [comment]
I don't know how it works for the linux / OS X solutions. But the windows one copies the binaries over to the output.

However, you certainly can run the tests. And if the results is green and NOT inconclusive/ignored, that should be fine.

**Comment by pblasucci** [comment]
@DanielFabian one quick change? Can we ignore `tests/TestResults.xml`? Committing it to the repo doesn't really provide much value, does it?

**Comment by DanielFabian** [comment]
ya sure, I'm not sure, why they were included in the first place.

**Comment by pblasucci** [comment]
I think I accidentally committed it ages ago.  Sorry... my bad :-(

**Comment by pblasucci** [comment]
Annnnd I just accidentally closed the PR. Whoops! Maybe I should step away from the keyboard before I screw something up.   ;-)

**Comment by haf** [comment]
Build works, then it freezes in unit tests.

```
Build succeeded.

Warnings:

/Users/h/dev/haf/fszmq/fszmq-osx.sln (Rebuild) ->
(Rebuild target) ->
/Users/h/dev/haf/fszmq/tests/fszmq.tests/fszmq.tests-osx.fsproj (Rebuild) ->
/Library/Frameworks/Mono.framework/Versions/4.4.1/lib/mono/4.5/Microsoft.FSharp.Targets (CoreCompile target) ->

	/Users/h/dev/haf/fszmq/tests/fszmq.tests/Miscellany.fs(53,54): warning FS0057: WARNING: Functionality in the Curve module requires more testing.. This warning can be disabled using '--nowarn:57' or '#nowarn "57"'.

	 1 Warning(s)
	 0 Error(s)

Time Elapsed 00:00:24.1648330
Finished Target: Build
Starting Target: CopyBinaries (==> Build)
Finished Target: CopyBinaries
Starting Target: RunTests (==> CopyBinaries)
/Users/h/dev/haf/fszmq/packages/NUnit.Runners/tools/nunit-console.exe "-nologo" "-noshadow" "-labels" "/Users/h/dev/haf/fszmq/tests/fszmq.tests/bin/Release/fszmq.tests.dll" "-xml:tests/TestResults.xml"
mono  /Users/h/dev/haf/fszmq/packages/NUnit.Runners/tools/nunit-console.exe "-nologo" "-noshadow" "-labels" "/Users/h/dev/haf/fszmq/tests/fszmq.tests/bin/Release/fszmq.tests.dll" "-xml:tests/TestResults.xml"
WARNING: The runtime version supported by this application is unavailable.
Using default runtime: v4.0.30319
ProcessModel: Default    DomainUsage: Single
Execution Runtime: mono-4.0
***** fszmq.tests.Curve.curveKeyPair generates 2 different evenly-divisible strings
Ok, passed 100 tests.
***** fszmq.tests.Message.after moving, target content should equal original content
Ok, passed 100 tests.
***** fszmq.tests.Message.both message should have the same content after copying
Ok, passed 100 tests.
***** fszmq.tests.Message.clone returns a new instance with the same data
Ok, passed 100 tests.
***** fszmq.tests.Message.copy requires two distinct message instances
***** fszmq.tests.Message.move requires two distinct message instances
***** fszmq.tests.Miscellany.libzmq version should be at least 4
***** fszmq.tests.Miscellany.recv throws TimeoutException if RCVTIMEO expires
***** fszmq.tests.Options.getting options does not throw: Events
***** fszmq.tests.Options.getting options does not throw: LastEndpointAddress
***** fszmq.tests.Options.getting options does not throw: MoreMessageFramesAvailable
***** fszmq.tests.Options.getting the security mechanism does not throw
***** fszmq.tests.Options.setting options and reading them are inverses: Affinity
***** fszmq.tests.Options.setting options and reading them are inverses: AuthenticationDomain
***** fszmq.tests.Options.setting options and reading them are inverses: Backlog
***** fszmq.tests.Options.setting options and reading them are inverses: CurveClient
***** fszmq.tests.Options.setting options and reading them are inverses: CurveServer
***** fszmq.tests.Options.setting options and reading them are inverses: GssapiClient
***** fszmq.tests.Options.setting options and reading them are inverses: GssapiClientUnencrypted
***** fszmq.tests.Options.setting options and reading them are inverses: GssapiServer
***** fszmq.tests.Options.setting options and reading them are inverses: GssapiServerUnencrypted
***** fszmq.tests.Options.setting options and reading them are inverses: HandshakeInterval
***** fszmq.tests.Options.setting options and reading them are inverses: Identity
***** fszmq.tests.Options.setting options and reading them are inverses: Immediate false
***** fszmq.tests.Options.setting options and reading them are inverses: Immediate true
***** fszmq.tests.Options.setting options and reading them are inverses: Ipv6 false
***** fszmq.tests.Options.setting options and reading them are inverses: Ipv6 true
***** fszmq.tests.Options.setting options and reading them are inverses: KeepLastMessageInQueue false
***** fszmq.tests.Options.setting options and reading them are inverses: KeepLastMessageInQueue true
***** fszmq.tests.Options.setting options and reading them are inverses: LingerDelay
***** fszmq.tests.Options.setting options and reading them are inverses: MaxMessageSize
***** fszmq.tests.Options.setting options and reading them are inverses: MaxReconnectInterval
***** fszmq.tests.Options.setting options and reading them are inverses: MulticastHops
***** fszmq.tests.Options.setting options and reading them are inverses: MulticastRecovery
***** fszmq.tests.Options.setting options and reading them are inverses: NullSecurity
***** fszmq.tests.Options.setting options and reading them are inverses: PlainClient
***** fszmq.tests.Options.setting options and reading them are inverses: PlainServer
***** fszmq.tests.Options.setting options and reading them are inverses: Rate
***** fszmq.tests.Options.setting options and reading them are inverses: ReceiveBuffer
***** fszmq.tests.Options.setting options and reading them are inverses: ReceiveQueue
***** fszmq.tests.Options.setting options and reading them are inverses: ReceiveTimeout
***** fszmq.tests.Options.setting options and reading them are inverses: ReconnectDelay
***** fszmq.tests.Options.setting options and reading them are inverses: SendBuffer
***** fszmq.tests.Options.setting options and reading them are inverses: SendQueue
***** fszmq.tests.Options.setting options and reading them are inverses: SendTimeout
***** fszmq.tests.Options.setting options and reading them are inverses: SocksProxy
***** fszmq.tests.Options.setting options and reading them are inverses: TcpKeepalive (Some false)
***** fszmq.tests.Options.setting options and reading them are inverses: TcpKeepalive (Some true)
***** fszmq.tests.Options.setting options and reading them are inverses: TcpKeepalive None
***** fszmq.tests.Options.setting options and reading them are inverses: TcpKeepaliveCount
***** fszmq.tests.Options.setting options and reading them are inverses: TcpKeepaliveIdle
***** fszmq.tests.Options.setting options and reading them are inverses: TcpKeepaliveInterval
***** fszmq.tests.Options.setting options and reading them are inverses: TypeOfService
***** fszmq.tests.Options.setting options does not throw: ConnectPeerId
***** fszmq.tests.Options.setting options does not throw: DoNoSilentlyDropMessages false
***** fszmq.tests.Options.setting options does not throw: DoNoSilentlyDropMessages true
***** fszmq.tests.Options.setting options does not throw: ProbeRouter false
***** fszmq.tests.Options.setting options does not throw: ProbeRouter true
***** fszmq.tests.Options.setting options does not throw: RelaxStrictAlternation false
***** fszmq.tests.Options.setting options does not throw: RelaxStrictAlternation true
***** fszmq.tests.Options.setting options does not throw: RequestCorrelation false
***** fszmq.tests.Options.setting options does not throw: RequestCorrelation true
***** fszmq.tests.Options.setting options does not throw: ResendDuplicateMessages false
***** fszmq.tests.Options.setting options does not throw: ResendDuplicateMessages true
***** fszmq.tests.Options.setting options does not throw: RouterHandover false
***** fszmq.tests.Options.setting options does not throw: RouterHandover true
***** fszmq.tests.Options.setting options does not throw: RouterMandatory false
***** fszmq.tests.Options.setting options does not throw: RouterMandatory true
***** fszmq.tests.Options.setting options does not throw: Subscribe "prefix"B
***** fszmq.tests.Options.setting options does not throw: Unsubscribe "prefix"B
***** fszmq.tests.Options.SocketType pattern should equal ZMQ.TYPE option
***** fszmq.tests.Termination.everything should shutdown cleanly
(1) running... cancel requested.
```

When using `==> ./configure --prefix=/usr/local/Cellar/zeromq/4.1.5 --with-pgm --with-libsodium` with `9055248834c3982c8d0a1412133149cbb54fb8f7  lib/zeromq/OSX/lib/libzmq.dylib`

**Comment by pblasucci** [comment]
Yeah... that's an unrelated test that should be disabled.

**Comment by DanielFabian** [comment]
cool!!! thanks for merging it in

**Comment by pblasucci** [comment]
No. Thank _YOU_ for all the hard work and awesome new API.  :-)

**Comment by DanielFabian** [comment]
thanks ;-)

**Pull request merged**